cdap-data-fabric/src/main/java/co/cask/cdap/spi/data/transaction/TransactionRunners.java

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.spi.data.transaction;
+
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Helper class for interacting with {@link TransactionRunner}
+ */
+public class TransactionRunners {
+
+  /**
+   * Executes the given {@link TxRunnable} using the given {@link TransactionRunner}.
+   *
+   * @param txRunner the {@link TransactionRunner} to use for txRunner execution.
+   * @param runnable the {@link TxRunnable} to be executed inside a transaction
+   * @param <X> exception type of propagate type
+   * @throws X if failed to execute the given {@link co.cask.cdap.api.TxRunnable} in a transaction.
+   * If the TransactionFailureException has a cause in it, the cause is thrown as-is if it is an instance of X.
+   * @throws RuntimeException if cause is not an instance of X. The cause is wrapped with {@link RuntimeException}
+   * if it is not already a {@link RuntimeException}.
+   */
```

Comment:
This is not clear to me. When I call this with a TxRunnable, and that TxRunnable throws A, then I would expect that:
- run() throws A if the runnable threw A
- run() throws TransactionException if the error is caused by the transaction system (deadlocks, conflicts, timeout, etc.)
- run() throws ExecutionException or something similar if run() throws something unexpected.

Reply:
This is same as our existing transaction api. There are different run() method which will let you specify the kind of exception the `TxRunnable` will throw, which is the generic type X here. If the TxRunnable throws X, X will be cause of the `TransactionException`, and the run() method will throw X as the exception. All the other exceptions are thrown as RuntimeException. I am unsure if we want to improve this as part of this spi, should I open a JIRA as improvement for future?

Reply:
future Jira is OK, but I really don't think we should simply propagate all exceptions as runtime exceptions.